p2p/src/interface/types.rs

```diff
@@ -27,6 +27,10 @@ pub struct ConnectedPeer {
 
     pub ban_score: u32,
 
+    pub user_agent: String,
```

Comment:
`UserAgent` does not implement `serde::Serialize`.

Reply:
About this and SemVer, the other comment. My problem with String is that String is made to be printable. If type has a conversion A -> ToString, it doesn't always mean that String -> A is the same value or is even possible. This is at least how Rust expects it. The condition A == from_string(to_string(A)) isn't always true.

Reply:
Sorry, I'm not sure what should be done here. Should I implement `Serialize` for `UserAgent` and `SemVer`?
Something like this:
```
impl serde::Serialize for UserAgent {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        self.to_string().serialize(serializer)
    }
}
```

Reply:
What I meant is, either create a type to cover that, or use `Vec<u8>`. But String looks like a bad choice in my opinion.

Reply:
Sorry, I still don't understand what should be done here.

`Vec<u8>` won't work, because RPC will then return the value as a binary array (`"user_agent":[77,105,110,116,108,97,121,101,114,67,111,114,101]`):
```
curl -XPOST -H"Content-Type: application/json" http://localhost:3030 -d '{"jsonrpc":"2.0","id":1,"method":"p2p_get_connected_peers","params":null}'
{"jsonrpc":"2.0","result":[{"peer_id":1,"address":"127.0.0.1:55694","inbound":true,"ban_score":0,"user_agent":[77,105,110,116,108,97,121,101,114,67,111,114,101],"version":"0.1.0","ping_wait":null,"ping_last":1,"ping_min":0}],"id":1}```

Reply:
I'm sorry, I seem to be confused about the purpose of this.

If this is strictly for RPC, then this is OK, and then it should be labeled or documented that this is strictly for RPC. My confusion here is that I thought this is going over the p2p interface, not the RPC interface. We should have a strict conversion from String to UserAgent where the `A == from_string(to_string(A))` is always true.

Apologies for the confusion.

Reply:
I see now. I added a comment.